diff --git a/build/BUILD.bazel b/build/BUILD.bazel
index e96a7a5aa..e26eb2d40 100644
--- a/build/BUILD.bazel
+++ b/build/BUILD.bazel
@@ -19,8 +19,7 @@ clib_deps = [
     kong_install(
         name = "install-%s" % get_workspace_name(k),
         src = k,
-        # only install openssl headers
-        exclude = [] if k in ("@openssl",) else ["include"],
+        exclude = ["include"],
         prefix = "kong/lib" if k in ("@passwdqc", "@snappy") else "kong",
         strip_path = "snappy" if k == "@snappy" else "",
     )
